src/ClassRegistrator_impl.hpp

```diff
@@ -0,0 +1,139 @@
+#include <string>
+
+#include "sexp.hpp"
+
+namespace { // Implementation detail, so it's in an anon namespace
+    template<typename Class>
+    struct ClassRegistratorHelpers {
+        template<typename FieldType>
+        static sexp field_get(sexp context, sexp self, long n, sexp this_ptr_sexp) {
+            auto this_ptr = static_cast<Class *>(sexp_cpointer_value(this_ptr_sexp));
+            auto member_ptr = static_cast<FieldType Class::**>(sexp_cpointer_value(sexp_opcode_data(self)));
+
+            Chibi chibi(context);
+            return chibi.make_from(this_ptr->**member_ptr);
+        }
+
+        template<typename FieldType>
+        static sexp field_set(sexp context, sexp self, long n, sexp this_ptr_sexp, sexp sexp_new_value) {
+            auto this_ptr = static_cast<Class *>(sexp_cpointer_value(this_ptr_sexp));
+            auto member_ptr = static_cast<FieldType Class::**>(sexp_cpointer_value(sexp_opcode_data(self)));
+
+            Chibi chibi(context);
+            return convert_or_exception(chibi, self, sexp_new_value, this_ptr->**member_ptr);
+        }
+
+        template<typename Return, typename Arg1>
+        static sexp call(sexp context, sexp self, long n, sexp this_ptr_sexp, sexp arg1_sexp) {
+            auto this_ptr = static_cast<Class *>(sexp_cpointer_value(this_ptr_sexp));
+            auto fnc_ptr = static_cast<Return (Class::**) (Arg1)>(sexp_cpointer_value(sexp_opcode_data(self))); // Extract the function pointer from opcode data
+
+            Chibi chibi(context);
+
+            Arg1 arg1;
+            if (sexp set_result = convert_or_exception(chibi, self, arg1_sexp, arg1); set_result != SEXP_UNDEF) {
+                return set_result;
+            }
+
+            Return res = (this_ptr->**fnc_ptr)(arg1);
+
+            return chibi.make_from(res);
+        }
+
+        template<typename Return, typename Arg1, typename Arg2>
+        static sexp call(sexp context, sexp self, long n, sexp this_ptr_sexp, sexp arg1_sexp, sexp arg2_sexp) {
+            auto this_ptr = static_cast<Class *>(sexp_cpointer_value(this_ptr_sexp));
+            auto fnc_ptr = static_cast<Return (Class::**) (Arg1, Arg2)>(sexp_cpointer_value(sexp_opcode_data(self)));
+
+            Chibi chibi(context);
+
+            Arg1 arg1;
+            if (sexp set_result = convert_or_exception(chibi, self, arg1_sexp, arg1); set_result != SEXP_UNDEF) {
+                return set_result;
+            }
+            Arg2 arg2;
+            if (sexp set_result = convert_or_exception(chibi, self, arg2_sexp, arg2); set_result != SEXP_UNDEF) {
+                return set_result;
+            }
+
+            Return res = (this_ptr->**fnc_ptr)(arg1, arg2);
+
+            return chibi.make_from(res);
+        }
+    private:
+        /** Convert from sexp to the expected value type and set to to \a set_result_to or return an exception.
```

Comment:
Documentation on private functions, when some public interface is not documented is a questionable thing.

Reply:
This function is documented because it's logic is complex and i'd like to remeber how it works.

Reply:
Ok.